### PR TITLE
Fix warnings when compiling for macOS Mojave

### PR DESCRIPTION
**Pull request by danzimm**
I was mildly annoyed by the warnings when compiling so I fixed them. Also added -Werror to avoid commits like this in the future.

### CODE REVIEW

**Comment by danzimm** [comment]
Before I dig into the failures, is -werror a welcome addition here?

**Comment by mqudsi** [comment]
@danzimm 

WRT 443303b670d45bb7662430937850303c5c74605e: 

Thanks for fixing the warnings under macOS. The code generally compiles without any compiler warnings on any of the supported platforms, this must be specific to your compiler (can you provide info on your C++ toolchain?). Much appreciated.

WRT af5846199d91deda7454270374bb9de57585b542:

`-Werror `should *never* be used by default in build scripts but should *always* be enabled on developers' machines. Perhaps we could enable it by default on Travis, but certainly this commit as-is is a big no-go.

_Note that there are even some (linker) warnings more-or-less by-design (see #4183)._

File: configure.ac [hunk 1]
```diff
@@ -220,7 +220,7 @@ CXXFLAGS="$CXXFLAGS -fno-exceptions"
 # Don't warn about missing field initializers, it has too many
 # false positives for code like `struct termios tmodes = {};`
 #
-CXXFLAGS="$CXXFLAGS -Wextra -Wno-missing-field-initializers"
+CXXFLAGS="$CXXFLAGS -Werror -Wextra -Wno-missing-field-initializers"
```
Comment (mqudsi) [inline configure.ac]: This isn't correct (see my comments in the discussion).

**Comment by danzimm** [comment]
@mqudsi I've removed the commit to add -Werror.

Re my toolchain: I just bought a new Macbook Air a day ago and installed the command line developer tools from Apple (the OS is Mojave). `clang --version` output:
```
Apple LLVM version 10.0.0 (clang-1000.10.44.4)
Target: x86_64-apple-darwin18.2.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

**Comment by faho** [comment]
>The code generally compiles without any compiler warnings on any of the supported platforms

(This is not entirely true. On recent-ish glibcs we get a warning about sys_errlist - see #4183)

**Comment by mqudsi** [comment]
I mentioned that in the same comment ;)

**Pull request merged**